File: problems/Minimum Insertion Steps to Make a String Palindrome/minInsertions.cs
public class Solution {
    public int MinInsertions(string s) {
        return s.Length - getLpsLength(s);
    }

    private int getLpsLength(String s) {
        int n = s.Length;
        int[,] dp = new int[n, n];

        for (int start = n - 1; 0 <= start; --start) {
            for (int end = start; n > end; ++end) {
                if (start == end) {
                    dp[start, end] = 1;
                } else if (s[start] == s[end]) {
                    dp[start, end] = 2 + dp[1 + start, end - 1];
                } else {
                    dp[start, end] = Math.Max(dp[1 + start, end], dp[start, end - 1]);
                }
            }
        }

        return dp[0, n - 1];
    }
}

File: problems/Minimum Insertion Steps to Make a String Palindrome/minInsertions.java
class Solution {
    public int minInsertions(String s) {
        return s.length() - getLpsLength(s);
    }

    private int getLpsLength(String s) {
        int n = s.length();
        Integer[][] dp = new Integer[n][n];

        return getLpsLengthTopDown(s, dp, 0, n - 1);
    }

    private int getLpsLengthTopDown(String s, Integer[][] dp, int start, int end) {
        if (start > end) {
            return 0;
        }

        if (start == end) {
            return 1;
        }

        if (null == dp[start][end]) {
            if (s.charAt(start) == s.charAt(end)) {
                dp[start][end] = 2 + getLpsLengthTopDown(s, dp, 1 + start, end - 1);
            } else {
                dp[start][end] = Math.max(getLpsLengthTopDown(s, dp, 1 + start, end), getLpsLengthTopDown(s, dp, start, end - 1));
            }
        }

        return dp[start][end];
    }
}
